### DISCUSSION TITLE
How to test App router and RSC with Jest

### DISCUSSION THREAD

**Discussion by Gpx** [vercel/next.js#49603] [App Router]
Hey there, I'm looking into migrating a Next.js app to the new App Router. Unfortunately, I'm hitting a roadblock. How do I test React Server Components in Jest?

Imagine I have this page:

```tsx
async function getUser() {
  const response = await fetch('/user')
  return response.json()
}

export default async function Page() {
  const user = await getUser()
  return <h1>Hello {user.name}!</h1>
}
```

How can I test this? Is it possible, or should I start migrating all my tests to Cypress or some other E2E library?

**Comment by ImBIOS** [comment]
for your example case, we should use e2e testing method.

but for unit, integration or component testing you can still use Jest, ref: https://stackoverflow.com/questions/75729282/testing-an-async-server-component-with-jest-in-next-13

**Reply by Gpx** [reply]
The solution linked is not working in React Testing Library. `render` will throw an error when given async components. See https://github.com/testing-library/react-testing-library/issues/1209

**Reply by nickserv** [reply]
That error is usually caused by using an older version of React. React 18.3 (canary) is required to render RSCs. Note that Next installs React stable locally but replaces it with canary in builds by default.

**Comment by nickserv** [comment]
You can test async components that don't use other RSC or app router features with React 18.3 (canary):
```jsx
import { render, screen } from '@testing-library/react'
import { Suspense } from 'react'
import Page from './page'

test('User page', async () => {
  render(
    <Suspense>
      <Page />
    </Suspense>
  )
  const heading = await screen.findByRole('heading')
  expect(heading).toHaveTextContent('Hello Gio')
})
```

You may want to use a [custom `render` function](https://testing-library.com/docs/react-testing-library/setup/#custom-render) to simplify test setup if your suite heavily relies on async components.

If you need other RSC (i.e. server actions) or app router (i.e. layouts) features you can use hard coding, mocks, or an e2e test framework until we figure out native RSC rendering in https://github.com/testing-library/react-testing-library/issues/1209.
